Implement json_escape jinja filter

Mierdin opened the pull request:
Example run:

```
(virtualenv)vagrant@st2dev:~/st2$ st2 run examples.json_escape_jinja_filter
..
id: 5944725b32ed3564e259663d
action.ref: examples.json_escape_jinja_filter
parameters: None
status: succeeded
start_timestamp: 2017-06-17T00:05:47.593411Z
end_timestamp: 2017-06-17T00:05:50.021274Z
+--------------------------+------------------------+------------+------------------------+-------------------------------+
| id                       | status                 | task       | action                 | start_timestamp               |
+--------------------------+------------------------+------------+------------------------+-------------------------------+
| 5944725b32ed3564a5ea1e49 | succeeded (2s elapsed) | post_alert | slack.chat.postMessage | Sat, 17 Jun 2017 00:05:47 UTC |
+--------------------------+------------------------+------------+------------------------+-------------------------------+
(virtualenv)vagrant@st2dev:~/st2$ st2 execution get 5944725b32ed3564a5ea1e49
id: 5944725b32ed3564a5ea1e49
status: succeeded (2s elapsed)
parameters:
  attachments: '[{"title": "FAILED - Random text from another action2 output """ which will break JSON and needs escaping", "color": "#808080"}]'
  channel: '#stackstorm'
  text: ' '
result:
  exit_code: 0
  result:
    channel: C029K4CBU
    message:
      attachments:
      - color: '808080'
        fallback: FAILED - Random text from another action2 output """ which will break JSON and needs escaping
        id: 1
        title: FAILED - Random text from another action2 output """ which will break JSON and needs escaping
      bot_id: B3ZN1TAVA
      text: ' '
      ts: '1497657948.159481'
      type: message
      user: U2WC08YJ2
    ok: true
    ts: '1497657948.159481'
  stderr: 'st2.actions.python.SlackAction: INFO     username=None&attachments=%5B%7B%22title%22%3A+%22FAILED+-+Random+text+from+another+action2+output+%5C%22%5C%22%5C%22+which+will+break+JSON+and+needs+escaping%22%2C+%22color%22%3A+%22%23808080%22%7D%5D&unfurl_links=None&link_names=None&unfurl_media=None&parse=None&token=xoxp-2324831254-98408304614-136363044439-a4cb0829393fc7ce5e29d9c5b94ded0b&text=+&icon_emoji=None&as_user=None&icon_url=None&channel=%23stackstorm

    '
  stdout: ''
```

Closes https://github.com/StackStorm/st2/issues/3447

bigmstone commented on st2common/st2common/jinja/filters/json_escape.py:
```diff
@@ -0,0 +1,49 @@
+        # These don't work yet, and I hate them because they are stupid.
+        # ("\f", r"\\f"),  # form feed (this doesn't work and I hate it)
+        # ("\r", r"\\r"),  # carriage return
+    ])
```
Could `re.escape` work here?

bigmstone replied:
Per slack convo use `json.dumps` instead.

lakshmi-kannan commented:
Looks good to me. I usually add examples in contrib so people can quickly try it out and we can refer to these in docs. 

https://github.com/StackStorm/st2/blob/master/contrib/examples/actions/chains/data_jinja_filter.yaml
https://github.com/StackStorm/st2/blob/master/contrib/examples/actions/chains/data_jinja_filter.yaml
https://github.com/StackStorm/st2/blob/master/contrib/examples/actions/chains/time_jinja_filter.yaml
https://github.com/StackStorm/st2/blob/master/contrib/examples/actions/chains/version_jinja_filter.yaml

This is also missing a CHANGELOG and you probably want to add info about the new filter in st2docs

armab commented on contrib/examples/actions/chains/json_escape_jinja_filter.yaml:
```diff
@@ -0,0 +1,11 @@
+     ref: "core.local"
+     parameters:
+        cmd: |
+          echo -E '[{"title":{{ escape_str | json_escape }}, "color": "#37C200"}]'
```
probably needs quotes `"title": "{{ escape_str | json_escape }}"`

lakshmi-kannan replied:
He has wrapper the whole thing in a ''. It's always a mess to include a shell command inside YAML.

lakshmi-kannan replied:
Can you also post output of example run in description?

Mierdin replied:
The same thing happened in my editor, but it works. I posted example in description

Mierdin replied:
Actually the reason it's on a separate line like that was to get around MY editor highlighting it that way, but apparently GH's works differently. Anyways it still works

armab replied:
Can you verify if this example works?
```yaml
---
vars:
    status: FAILED
    description: |
      Random text from another action output """ which will break JSON and needs escaping

chain:
  -  name: post_alert
     ref: slack.chat.postMessage
     parameters:
       text: " "
       channel: "#stackstorm"
       attachments: '[{"title": "{{ status }} - {{ description|json_escape }}", "color": "#808080"}]'
```

Mierdin replied:
Looks like it executes but no attachments show up in slack for some reason:

```
(virtualenv)vagrant@st2dev:~/st2$ st2 run examples.json_escape_jinja_filter
..
id: 5944522032ed3519c91ded18
action.ref: examples.json_escape_jinja_filter
parameters: None
status: succeeded
start_timestamp: 2017-06-16T21:48:16.243673Z
end_timestamp: 2017-06-16T21:48:18.664423Z
+--------------------------+------------------------+------------+------------------------+----------------------------+
| id                       | status                 | task       | action                 | start_timestamp            |
+--------------------------+------------------------+------------+------------------------+----------------------------+
| 5944522032ed35198b006ff0 | succeeded (2s elapsed) | post_alert | slack.chat.postMessage | Fri, 16 Jun 2017 21:48:16  |
|                          |                        |            |                        | UTC                        |
+--------------------------+------------------------+------------+------------------------+----------------------------+
(virtualenv)vagrant@st2dev:~/st2$ st2 execution get 5944522032ed35198b006ff0
id: 5944522032ed35198b006ff0
status: succeeded (2s elapsed)
parameters:
  attachments: '[{"title": "FAILED - "Random text from another action output """ which will break JSON and needs escaping"", "color": "#808080"}]'
  channel: '#stackstorm'
  text: ' '
result:
  exit_code: 0
  result:
    channel: C029K4CBU
    message:
      bot_id: B3ZN1TAVA
      text: ' '
      ts: '1497649697.891337'
      type: message
      user: U2WC08YJ2
    ok: true
    ts: '1497649697.891337'
  stderr: 'st2.actions.python.SlackAction: INFO     username=None&attachments=%5B%7B%22title%22%3A+%22FAILED+-+%22Random+text+from+another+action+output+%5C%22%5C%22%5C%22+which+will+break+JSON+and+needs+escaping%22%22%2C+%22color%22%3A+%22%23808080%22%7D%5D&unfurl_links=None&link_names=None&unfurl_media=None&parse=None&token=xoxp-2324831254-98408304614-136363044439-a4cb0829393fc7ce5e29d9c5b94ded0b&text=+&icon_emoji=None&as_user=None&icon_url=None&channel=%23stackstorm

    '
  stdout: ''
```

Mierdin replied:
Since I'm now using `json.dumps` I think the input may need to be JSON, so that example wouldn't work. Let me tinker a bit more

armab commented on CHANGELOG.rst:
```diff
@@ -4,6 +4,8 @@ Changelog
 in development
 --------------
 
+* Add json_ecsape Jinja filter for escaping JSON strings
```
Would be nice to reference #3447 or #3480 in the message, as well as include the type of record: is it improvement, new feature, bugfix.
See previous messages in the changelog for consistency.

Mierdin replied:
Ack that, done in 685c55b23

Mierdin commented:
As mentioned on Slack, I was reminded that `json.dumps` adds a double quote to the ends of the actual string value (as it should for it's primary use case - but for our use case it's not wanted). So I added a `.strip('"')`. Also updated the unit tests (which I wish I had looked at earlier, since it's pretty obvious there what the problem was. Just wasn't super obvious in the previous example I was using. Thanks @armab for a more realistic example.

bigmstone commented:
Would `strip(")` also remove desired quotes? How about `output[1:-1]`?

Mierdin commented:
What do you mean desired quotes?

On Jun 16, 2017 7:20 PM, "Matthew Stone" <notifications@github.com> wrote:

> Would strip(") also remove desired quotes? How about output[1:-1]?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/StackStorm/st2/pull/3480#issuecomment-309187014>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AECM--If8NHo2VJs0Kozd7oNadwBDlzcks5sEzf0gaJpZM4N6xLF>
> .
>

armab commented:
Looks better :+1:, and I agree with @bigmstone's note about the quotes.
Would be safer to remove first and last double quote in that string.

humbert73 commented:
```python
def json_escape(value):
    return json.dumps(value)
```
Seems to be suficient.
In my case I trie to import string with EOL characters and with just `json.dumps(value)` it's working but not with `json.dumps(value).strip('"')`

moreauxh commented:
```python
def json_escape(value):
    return json.dumps(value)
```
Seems to be suficient.
In my case I trie to import string with EOL characters and with just `json.dumps(value)` it's working but not with `json.dumps(value).strip('"')`

bigmstone commented:
@armab @Mierdin 

From `strip` docstring:
> Docstring:
   S.strip([chars]) -> string or unicode
   
   Return a copy of the string S with leading and trailing
   whitespace removed.
   If chars is given and not None, remove characters in chars instead.
   If chars is unicode, S will be converted to unicode before stripping
   Type:      builtin_function_or_method


So I assume this method is actually correct.

Testing this would prove it to be:
```
In [1]: s = '"This is my string. But, gasp, it has a " "in it"! And I do not want it removed!"'

In [2]: s.strip('"')
Out[2]: 'This is my string. But, gasp, it has a " "in it"! And I do not want it removed!'
```

armab commented:
Thanks @bigmstone for clarifying. Looks good to me then :+1:

armab commented:
@moreauxh Can you show exact string which won't work?  So we could include it as test case.

moreauxh commented:
I'm in a case where I use this:
```yaml
---
    chain:
        -
            name: "list_issues"
            ref: "github.list_issues"
            parameters:
                repo: "{{ repo }}"
                user: "{{ user }}"
            publish:
                formated_results: "{
                    {% for result in  list_issues.result %}
                        '{{ result.id }}': {
                            'login':       {'value': '{{ result.author.login }}' },
                            'name':        {'value': '{{ result.author.name }}' },
                            'title':       {'value': '{{ result.title }}' },
                            'description': {'value': {{ result.body|json_escape }} }
                        },
                    {% endfor %}
                }"
            on-success: "create_artifacts"
            on-failure: "fail"
        -
            name: "create_artifacts"
            ref: "tuleap.create_artifacts"
            parameters:
                tracker_id: "{{ tracker_id }}"
                values_by_field_by_artifact: "{{ formated_results }}"
```
And if the  `result.body` is a text with special characters like:
```
Some text.'

test "special" charactere.
```
My json will be break for the step 2 of my workflow.

Error Message:
```
error: Failed to cast value "{ '194603594': { 'login':       {'value': 'humbert73' }, 'name':        {'value': 'Humbert Moreaux' }, 'title':       {'value': 'Remove the broken Google plugin of PHPWiki' }, 'description': {'value': When I create my account, an error message is displayed.'

test "special" charactere.  } }  }" (type: unicode) for parameter "values_by_field_by_artifact" of type "object": invalid syntax (<unknown>, line 1). Perhaphs the value is of an invalid type?
```

armab commented:
@moreauxh Try to wrap strings not with single qoutes, but with double qoutes, which is the json spec http://www.json.org/

Something like this: 

```yaml
---
                formated_results: |
                  {
                    {% for result in  list_issues.result %}
                        "{{ result.id }}": {
                            "login": {"value": "{{ result.author.login }}"},
                            "name": {"value": "{{ result.author.name }}"},
                            "title": {"value": "{{ result.title }}"},
                            "description": {"value": "{{ result.body|json_escape }}"}
                        },
                    {% endfor %}
                  }
```

moreauxh commented:
Thanks it's working

armab commented:
Seems everyone is happy, + :100: 
Just a reminder that this change will need respective `st2docs` update.

Mierdin commented:
@armab Yup, thanks - I was waiting for this and https://github.com/StackStorm/st2/pull/3482 to merge then I was going to re-work the whole filters page, since it's in some need of it
